From 001442b33574d450a5fe6cb8d7f9528cf77062ac Mon Sep 17 00:00:00 2001
From: Chao Guo <chao.guo@nxp.com>
Date: Sat, 12 Oct 2024 11:13:31 +0900
Subject: [PATCH 85/91] LF-13361 glcolorconvert: Avoid copying texture to
 another texture bound to dmabuf on GLES2 platforms

In GLES2 platform, out size and mem size may be different, causing memory copy after converting.
For DMA memory, skip copying, because glCopyTexImage2D() cannot copy texture to dmabuf.

Upstream-Status: Inappropriate [i.MX specific]

Signed-off-by: Chao Guo <chao.guo@nxp.com>
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 gst-libs/gst/gl/gstglcolorconvert.c | 18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/gst-libs/gst/gl/gstglcolorconvert.c b/gst-libs/gst/gl/gstglcolorconvert.c
index e1d8b13..59e8132 100644
--- a/gst-libs/gst/gl/gstglcolorconvert.c
+++ b/gst-libs/gst/gl/gstglcolorconvert.c
@@ -31,6 +31,10 @@
 #include "gstglfuncs.h"
 #include "gstglsl_private.h"
 
+#if GST_GL_HAVE_IONDMA || GST_GL_HAVE_DMABUFHEAPS
+#include <gst/gl/gstglmemorydma.h>
+#endif
+
 /**
  * SECTION:gstglcolorconvert
  * @title: GstGLColorConvert
@@ -2740,12 +2744,22 @@ _do_convert_one_view (GstGLContext * context, GstGLColorConvert * convert,
   gint i, j = 0;
   const gint in_plane_offset = view_num * c_info->in_n_textures;
   const gint out_plane_offset = view_num * c_info->out_n_textures;
+  gboolean can_copy = TRUE;
 
   out_width = GST_VIDEO_INFO_WIDTH (&convert->out_info);
   out_height = GST_VIDEO_INFO_HEIGHT (&convert->out_info);
   in_width = GST_VIDEO_INFO_WIDTH (&convert->in_info);
   in_height = GST_VIDEO_INFO_HEIGHT (&convert->in_info);
 
+#if GST_GL_HAVE_IONDMA || GST_GL_HAVE_DMABUFHEAPS
+  /* In GLES2 platform, out size and mem size may be different,
+   * causing memory copy after converting. For DMA memory, skip
+   * copying, because glCopyTexImage2D() cannot copy texture to dmabuf. */
+  can_copy = USING_GLES3 (context) ||
+      !gst_is_gl_memory_dma ((GstMemory *) gst_buffer_peek_memory (convert->outbuf,
+                                                                   out_plane_offset));
+#endif
+
   for (i = 0; i < c_info->in_n_textures; i++) {
     convert->priv->in_tex[i] =
         (GstGLMemory *) gst_buffer_peek_memory (convert->inbuf,
@@ -2796,7 +2810,7 @@ _do_convert_one_view (GstGLContext * context, GstGLColorConvert * convert,
 
     if (out_tex->tex_format == GST_GL_LUMINANCE
         || out_tex->tex_format == GST_GL_LUMINANCE_ALPHA
-        || out_width != mem_width || out_height != mem_height) {
+        || ((out_width != mem_width || out_height != mem_height) && can_copy)) {
       /* Luminance formats are not color renderable */
       /* rendering to a framebuffer only renders the intersection of all
        * the attachments i.e. the smallest attachment size */
@@ -2868,7 +2882,7 @@ out:
 
     if (out_tex->tex_format == GST_GL_LUMINANCE
         || out_tex->tex_format == GST_GL_LUMINANCE_ALPHA
-        || out_width != mem_width || out_height != mem_height) {
+        || ((out_width != mem_width || out_height != mem_height) && can_copy)) {
       GstMapInfo to_info, from_info;
 
       if (!gst_memory_map ((GstMemory *) convert->priv->out_tex[j], &from_info,
-- 
2.43.0

